packages/structures/src/bitfields/BitField.ts

```diff
@@ -0,0 +1,205 @@
+import type { EnumLike, RecursiveReadonlyArray } from '../utils/types';
+
+/**
+ * Data that can be resolved to give a bitfield. This can be:
+ * A bit number (this can be a number literal or a value taken from {@link BitField.Flags})
+ * A string bit number
+ * An instance of BitField
+ * An Array of BitFieldResolvable
+ */
+export type BitFieldResolvable<Flags extends string, Type extends bigint | number> =
+	| Flags
+	| Readonly<BitField<Flags, Type>>
+	| RecursiveReadonlyArray<Flags | Readonly<BitField<Flags, Type>> | Type | `${bigint}`>
+	| Type
+	| `${bigint}`;
+
+/**
+ * Data structure that makes it easy to interact with a bitfield.
+ */
+export abstract class BitField<Flags extends string, Type extends bigint | number = number> {
+	/**
+	 * Numeric bitfield flags.
+	 * <info>Defined in extension classes</info>
+	 */
+	public static readonly Flags: EnumLike<unknown, bigint | number> = {};
+
+	protected static readonly DefaultBit: bigint | number = 0;
```

Comment:
Random thought, cc @didinele, what if we just always enforce bigints for bitfields? And we convert to string/number in djs as needed? Makes it far easier to work with in the future if an enum moves from numbers to bigints

Reply:
Well, I was going to initially say, is there a reason why we want `number`s at all? Is the BigInt overhead *that big* for numbers that would fit in `number` anyway?